File: app/src/main/java/com/yibao/music/activity/SearchActivity.java
package com.yibao.music.activity;

import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.text.Editable;
import android.view.View;
import android.view.inputmethod.InputMethodManager;

import androidx.annotation.Nullable;
import androidx.viewpager2.widget.ViewPager2;

import com.jakewharton.rxbinding2.view.RxView;
import com.yibao.music.R;
import com.yibao.music.adapter.SearchPagerAdapter;
import com.yibao.music.base.BaseTransitionActivity;
import com.yibao.music.base.listener.OnFlowLayoutClickListener;
import com.yibao.music.base.listener.OnMusicItemClickListener;
import com.yibao.music.base.listener.TextChangedListener;
import com.yibao.music.databinding.ActivitySearchBinding;
import com.yibao.music.model.MusicBean;
import com.yibao.music.model.MusicLyricBean;
import com.yibao.music.model.SearchCategoryBean;
import com.yibao.music.service.MusicPlayService;
import com.yibao.music.util.ColorUtil;
import com.yibao.music.util.Constants;
import com.yibao.music.util.FileUtil;
import com.yibao.music.util.LyricsUtil;
import com.yibao.music.util.SoftKeybordUtil;
import com.yibao.music.util.TitleArtistUtil;

import java.util.List;
import java.util.concurrent.TimeUnit;

import io.reactivex.Observable;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.Disposable;
import io.reactivex.schedulers.Schedulers;

/**
 * @author lsp
 */
public class SearchActivity extends BaseTransitionActivity implements OnMusicItemClickListener, OnFlowLayoutClickListener, View.OnClickListener {
    private ActivitySearchBinding mBinding;
    private MusicBean mMusicBean;
    private MusicPlayService.AudioBinder audioBinder;
    private int lyricsFlag;
    private InputMethodManager mInputMethodManager;
    private Disposable mDisposableSoftKeyboard;
    private String mSearchCondition;
    private int currentCategoryPosition = 0;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_search);
        mBinding = ActivitySearchBinding.inflate(getLayoutInflater());
        setContentView(mBinding.getRoot());
        init();
        initListener();
    }

    @Override
    protected void updateLyricsView(boolean lyricsOK, String downMsg) {
        if (lyricsOK) {
            updateLyric();
        }
    }

    private void init() {
        int pageType = getIntent().getIntExtra("pageType", 0);
        audioBinder = MusicActivity.getAudioBinder();
        mBinding.smartisanControlBar.setPbColorAndPreBtnGone();
        SearchPagerAdapter pagerAdapter;
        if (pageType > Constants.NUMBER_ZERO) {
            mMusicBean = getIntent().getParcelableExtra("musicBean");
            mBinding.editSearch.setText(mMusicBean.getArtist());
            mBinding.editSearch.setSelection(mMusicBean.getArtist().length());
            mBinding.searchCategoryRoot.getRoot().setVisibility(View.VISIBLE);
            // ViewPager
            pagerAdapter = new SearchPagerAdapter(this, mMusicBean.getArtist());
            switchListCategory(3);
            setMusicInfo(mMusicBean);
            mBinding.ivEditClear.setVisibility(View.VISIBLE);
        } else {
            pagerAdapter = new SearchPagerAdapter(this, null);
            // 主动弹出键盘
            mInputMethodManager = (InputMethodManager) getSystemService(Context.INPUT_METHOD_SERVICE);
            mDisposableSoftKeyboard = Observable.timer(50, TimeUnit.MILLISECONDS)
                    .subscribe(aLong -> SoftKeybordUtil.showAndHintSoftInput(mInputMethodManager, 2, InputMethodManager.SHOW_FORCED));
        }

        mBinding.vpSearch.setAdapter(pagerAdapter);
        mBinding.vpSearch.registerOnPageChangeCallback(new ViewPager2.OnPageChangeCallback() {
            @Override
            public void onPageSelected(int position) {
                switchListCategory(position);
            }
        });
    }


    @Override
    protected void onResume() {
        super.onResume();
        if (audioBinder != null) {
            mMusicBean = audioBinder.getMusicBean();
            setMusicInfo(mMusicBean);
            checkCurrentSongIsFavorite(mMusicBean, null, mBinding.smartisanControlBar);
            mBinding.smartisanControlBar.updatePlayBtnStatus(audioBinder.isPlaying());
            mBinding.smartisanControlBar.animatorOnResume(audioBinder.isPlaying());
            updateLyric();
            setDuration();
        }
        mCompositeDisposable.add(RxView.clicks(mBinding.smartisanControlBar)
                .throttleFirst(1, TimeUnit.SECONDS)
                .subscribe(o -> startPlayActivity()));

    }


    @Override
    protected void updateCurrentPlayInfo(MusicBean musicItem) {
        mMusicBean = musicItem;
        SearchActivity.this.setMusicInfo(musicItem);
        mBinding.smartisanControlBar.initAnimation();
        if (audioBinder != null) {
            setDuration();
            SearchActivity.this.checkCurrentSongIsFavorite(mMusicBean, null, mBinding.smartisanControlBar);
            mBinding.smartisanControlBar.updatePlayBtnStatus(audioBinder.isPlaying());
            updateLyric();
        }
    }

    @Override
    protected void updateCurrentPlayProgress() {
        if (audioBinder != null) {
            mBinding.smartisanControlBar.setSongProgress(audioBinder.getProgress());
        }
    }

    private void setDuration() {
        int duration = audioBinder.getDuration();
        mBinding.smartisanControlBar.setMaxProgress(duration);
    }

    private void setMusicInfo(MusicBean musicItem) {
        if (musicItem != null) {
            mBinding.smartisanControlBar.setVisibility(View.VISIBLE);
            musicItem = TitleArtistUtil.getMusicBean(musicItem);
            mBinding.smartisanControlBar.setSongName(musicItem.getTitle());
            mBinding.smartisanControlBar.setSingerName(musicItem.getArtist());
            mBinding.smartisanControlBar.setAlbulmUrl(FileUtil.getAlbumUrl(musicItem, 1));
        }
        if (audioBinder != null) {
            mBinding.smartisanControlBar.updatePlayBtnStatus(audioBinder.isPlaying());
            mBinding.smartisanControlBar.initAnimation();
        } else {
            mBinding.smartisanControlBar.setVisibility(View.GONE);
        }
    }


    private void switchListCategory(int position) {
        currentCategoryPosition = position;
        mBinding.vpSearch.setCurrentItem(position, false);
        mBus.post(new SearchCategoryBean(getDataFlag(), mSearchCondition));
        switch (position) {
            case 0:
                setAllCategoryNotNormal();
                mBinding.searchCategoryRoot.tvSearchAll.setTextColor(ColorUtil.wihtle);
                mBinding.searchCategoryRoot.tvSearchAll.setBackgroundResource(R.drawable.btn_category_songname_down_selector);
                break;
            case 1:
                setAllCategoryNotNormal();
                mBinding.searchCategoryRoot.tvSearchSong.setTextColor(ColorUtil.wihtle);
                mBinding.searchCategoryRoot.tvSearchSong.setBackgroundResource(R.drawable.btn_category_score_down_selector);
                break;
            case 2:
                setAllCategoryNotNormal();
                mBinding.searchCategoryRoot.tvSearchAlbum.setTextColor(ColorUtil.wihtle);
                mBinding.searchCategoryRoot.tvSearchAlbum.setBackgroundResource(R.drawable.btn_category_score_down_selector);
                break;
            case 3:
                setAllCategoryNotNormal();
                mBinding.searchCategoryRoot.tvSearchArtist.setBackgroundResource(R.drawable.btn_category_views_down_selector);
                mBinding.searchCategoryRoot.tvSearchArtist.setTextColor(ColorUtil.wihtle);
                break;
            default:
                break;
        }
    }

    private void setAllCategoryNotNormal() {
        mBinding.searchCategoryRoot.tvSearchAll.setTextColor(ColorUtil.textName);
        mBinding.searchCategoryRoot.tvSearchAll.setBackgroundResource(R.drawable.btn_category_songname_selector);
        mBinding.searchCategoryRoot.tvSearchSong.setTextColor(ColorUtil.textName);
        mBinding.searchCategoryRoot.tvSearchSong.setBackgroundResource(R.drawable.btn_category_score_selector);
        mBinding.searchCategoryRoot.tvSearchAlbum.setTextColor(ColorUtil.textName);
        mBinding.searchCategoryRoot.tvSearchAlbum.setBackgroundResource(R.drawable.btn_category_score_selector);
        mBinding.searchCategoryRoot.tvSearchArtist.setTextColor(ColorUtil.textName);
        mBinding.searchCategoryRoot.tvSearchArtist.setBackgroundResource(R.drawable.btn_category_views_selector);
    }

    @Override
    protected void refreshBtnAndNotify(int playStatus) {
        switch (playStatus) {
            case 0:
                mBinding.smartisanControlBar.animatorOnResume(audioBinder.isPlaying());
                mBinding.smartisanControlBar.updatePlayBtnStatus(audioBinder.isPlaying());
                break;
            case 1:
                checkCurrentSongIsFavorite(mMusicBean, null, mBinding.smartisanControlBar);
                break;
            case 2:
                mBinding.smartisanControlBar.updatePlayBtnStatus(audioBinder.isPlaying());
                mBinding.smartisanControlBar.animatorOnPause();
                break;
            default:
                break;
        }
    }

    private int getDataFlag() {
        return mSearchCondition != null ? 1 : currentCategoryPosition == 0 || currentCategoryPosition == 1 ?
                Constants.NUMBER_THREE : currentCategoryPosition == 2 ? 2 : currentCategoryPosition == 3 ? 1 : 4;
    }


    private void switchPlayState() {
        if (audioBinder.isPlaying()) {
            audioBinder.pause();
        } else {
            audioBinder.start();
        }
        mBinding.smartisanControlBar.animatorOnResume(audioBinder.isPlaying());
        mBinding.smartisanControlBar.updatePlayBtnStatus(audioBinder.isPlaying());
    }

    private void initListener() {
        mBinding.editSearch.addTextChangedListener(new TextChangedListener() {
            @Override
            public void afterTextChanged(Editable s) {
                mSearchCondition = s.toString();
                boolean conditionOk = !Constants.NULL_STRING.equals(mSearchCondition) && mSearchCondition.length() > 0;
                mBus.post(new SearchCategoryBean(conditionOk ? getDataFlag() : Constants.NUMBER_TEN, mSearchCondition));
                mBinding.ivEditClear.setVisibility(conditionOk ? View.VISIBLE : View.GONE);
                mBinding.searchCategoryRoot.getRoot().setVisibility(conditionOk ? View.VISIBLE : View.GONE);
            }
        });
        mBinding.smartisanControlBar.setClickListener(clickFlag -> {
            switch (clickFlag) {
                case Constants.NUMBER_ONE:
                    audioBinder.updataFavorite();
                    checkCurrentSongIsFavorite(mMusicBean, null, mBinding.smartisanControlBar);
                    break;
                case Constants.NUMBER_TWO:
                    audioBinder.playPre();
                    break;
                case Constants.NUMBER_THREE:
                    switchPlayState();
                    break;
                case Constants.NUMBER_FOUR:
                    audioBinder.playNext();
                    break;
                default:
                    break;
            }
        });

        mBinding.tvSearchCancel.setOnClickListener(this);
        mBinding.ivEditClear.setOnClickListener(this);
        mBinding.searchCategoryRoot.tvSearchAll.setOnClickListener(this);
        mBinding.searchCategoryRoot.tvSearchSong.setOnClickListener(this);
        mBinding.searchCategoryRoot.tvSearchArtist.setOnClickListener(this);
        mBinding.searchCategoryRoot.tvSearchAlbum.setOnClickListener(this);
    }


    @Override
    public void onClick(View v) {
        int id = v.getId();
        if (id == R.id.tv_search_cancel) {
            SoftKeybordUtil.showAndHintSoftInput(mInputMethodManager, 1, InputMethodManager.RESULT_UNCHANGED_SHOWN);
            finish();
        } else if (id == R.id.iv_edit_clear) {
            mBinding.editSearch.setText(null);
            findViewById(R.id.search_category_root).setVisibility(View.GONE);
            mBus.post(new SearchCategoryBean(Constants.NUMBER_NINE, null));
        } else if (id == R.id.tv_search_all) {
            switchListCategory(0);
        } else if (id == R.id.tv_search_song) {
            switchListCategory(1);
        } else if (id == R.id.tv_search_album) {
            switchListCategory(2);
        } else if (id == R.id.tv_search_artist) {
            switchListCategory(3);
        }
    }

    @Override
    public void startMusicServiceFlag(int position, int sortFlag, int dataFlag, String queryFlag) {
        Intent intent = new Intent(this, MusicPlayService.class);
        intent.putExtra("sortFlag", sortFlag);
        intent.putExtra("dataFlag", dataFlag);
        intent.putExtra("queryFlag", queryFlag);
        intent.putExtra("position", position);
        startService(intent);
    }

    @Override
    public void startMusicService(int position) {

    }

    @Override
    public void onOpenMusicPlayDialogFag() {

    }

    @Override
    protected void onPause() {
        super.onPause();
        mBinding.smartisanControlBar.animatorOnPause();
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        if (mDisposableSoftKeyboard != null) {
            mDisposableSoftKeyboard.dispose();
            mDisposableSoftKeyboard = null;
        }
        if (audioBinder != null) {
            audioBinder = null;
        }
    }

    private void updateLyric() {
        List<MusicLyricBean> lyricList = LyricsUtil.getLyricList(mMusicBean);
        disposableQqLyric();
        if (mQqLyricsDisposable == null) {
            mQqLyricsDisposable = Observable.interval(0, 2800, TimeUnit.MICROSECONDS)
                    .subscribeOn(Schedulers.io())
                    .observeOn(AndroidSchedulers.mainThread())
                    .subscribe(musicBeanList -> {
                        if (lyricList.size() > 1 && lyricsFlag < lyricList.size()) {
                            //通过集合，播放过的歌词就从集合中删除
                            MusicLyricBean lyrBean = lyricList.get(lyricsFlag);
                            String content = lyrBean.getContent();
                            int progress = audioBinder.getProgress();
                            int startTime = lyrBean.getStartTime();
                            if (progress > startTime) {
                                mBinding.smartisanControlBar.setSingerName(content);
                                lyricsFlag++;
                            }
                        }
                    });
        }

    }

    @Override
    public void click(String songName) {
        mBinding.editSearch.setText(songName);
        mBinding.editSearch.setSelection(songName.length());

    }

}

File: app/src/main/java/com/yibao/music/activity/SearchActivity.kt
package com.yibao.music.activity

import android.content.Intent
import android.text.Editable
import android.view.View
import android.view.inputmethod.InputMethodManager
import com.yibao.music.R
import com.yibao.music.adapter.DetailsViewAdapter
import com.yibao.music.base.bindings.BaseBindingActivity
import com.yibao.music.base.bindings.BaseBindingAdapter
import com.yibao.music.base.listener.OnFlowLayoutClickListener
import com.yibao.music.base.listener.OnMusicItemClickListener
import com.yibao.music.base.listener.TextChangedListener
import com.yibao.music.databinding.ActivitySearchBinding
import com.yibao.music.fragment.dialogfrag.MoreMenuBottomDialog
import com.yibao.music.model.MusicBean
import com.yibao.music.service.MusicPlayService
import com.yibao.music.service.MusicPlayService.AudioBinder
import com.yibao.music.util.ColorUtil
import com.yibao.music.util.Constant
import com.yibao.music.util.LogUtil
import com.yibao.music.util.LyricsUtil
import com.yibao.music.util.SoftKeybordUtil
import com.yibao.music.util.TitleArtistUtil
import com.yibao.music.viewmodel.SearchViewModel
import io.reactivex.Observable
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.schedulers.Schedulers
import java.util.concurrent.TimeUnit

/**
 * @author lsp
 */
class SearchActivity : BaseBindingActivity<ActivitySearchBinding>(), OnMusicItemClickListener,
    OnFlowLayoutClickListener, View.OnClickListener {
    private val mViewModel: SearchViewModel by lazy { gets(SearchViewModel::class.java) }
    private var mMusicBean: MusicBean? = null
    private var audioBinder: AudioBinder? = null
    private var lyricsFlag = 0
    private var mInputMethodManager: InputMethodManager? = null
    private var mAdapter: DetailsViewAdapter? = null

    // 默认为 11 ，按歌曲名搜索。
    private var mPosition = 11
    override fun initView() {
        initRecyclerView(mBinding.recyclerSearch)
    }

    override fun initData() {
        // pageType 0 toolbar上的搜索，1 播放界面点击歌曲名，直接以歌手搜索。
        val pageType = intent.getIntExtra(Constant.PAGE_TYPE, 0)
        LogUtil.d(TAG, "PageType   $pageType")
        audioBinder = MainActivity.audioBinder
        mBinding.smartisanControlBar.setPbColorAndPreBtnGone()
        // 从PlayActivity过来的,按歌手搜索。
        if (pageType > Constant.NUMBER_ZERO) {
            mMusicBean = intent.getParcelableExtra(Constant.MUSIC_BEAN)

            mBinding.editSearch.setText(mMusicBean!!.artist)
            switchListCategory(Constant.SEARCH_SONG_FOR_ARTIST)

        } else {
            // 加载历史记录
            mViewModel.getHistory()

            // 从Toolbar上过来的，主动弹出键盘
//            mInputMethodManager =
//                getSystemService(Context.INPUT_METHOD_SERVICE) as InputMethodManager
//            Timer().schedule(timerTask {
//                SoftKeybordUtil.showAndHintSoftInput(
//                    mInputMethodManager,
//                    2,
//                    InputMethodManager.SHOW_FORCED
//                )
//            }, 100)

        }
    }

    override fun onResume() {
        super.onResume()
        // 搜索结果
        mViewModel.searchViewModel.observe(this) { musicList ->
            setData(musicList)
        }
        mViewModel.historyViewModel.observe(this) { historyList ->
            mBinding.flowlayout.setData(historyList)
            mBinding.flowlayout.setItemClickListener { _, bean ->
//                mPosition = 2
                mBinding.editSearch.setText(bean.searchContent)
                mBinding.editSearch.setSelection(bean.searchContent.length)
            }
        }
        if (audioBinder != null) {
            mMusicBean = audioBinder!!.musicBean
            setMusicInfo(mMusicBean)
            checkCurrentSongIsFavorite(mMusicBean, null, mBinding.smartisanControlBar)
            mBinding.smartisanControlBar.updatePlayBtnStatus(audioBinder!!.isPlaying)
            mBinding.smartisanControlBar.animatorOnResume(audioBinder!!.isPlaying)
            updateLyric()
            setDuration()
        }

    }

    private fun setData(musicList: List<MusicBean>) {
        LogUtil.d(TAG, "当前页面标识   $mPosition")
        if (musicList.isEmpty()) {
            mBinding.tvNoSearchResult.visibility = View.VISIBLE
            mBinding.recyclerSearch.visibility = View.GONE
        } else {
            mBinding.tvNoSearchResult.visibility = View.GONE
            mBinding.recyclerSearch.visibility = View.VISIBLE
            val editCondition = mBinding.editSearch.text.toString().trim()
            // 列表数据
            mAdapter = DetailsViewAdapter(this, musicList, mPosition, editCondition)
            mBinding.recyclerSearch.adapter = mAdapter
            mAdapter!!.setOnItemMenuListener(object :
                BaseBindingAdapter.OnOpenItemMoreMenuListener {
                override fun openClickMoreMenu(position: Int, musicBean: MusicBean) {
                    // 关闭键盘
                    SoftKeybordUtil.showAndHintSoftInput(
                        mInputMethodManager, 1, InputMethodManager.SHOW_FORCED
                    )
                    MoreMenuBottomDialog.newInstance(
                        musicBean, position, false, false
                    ).getBottomDialog(this@SearchActivity)

                }
            })

        }

    }

    override fun initListener() {
        mBinding.tvSearchCancel.setOnClickListener(this)
        mBinding.ivEditClear.setOnClickListener(this)
        mBinding.searchCategoryRoot.tvSearchAll.setOnClickListener(this)
        mBinding.searchCategoryRoot.tvSearchSong.setOnClickListener(this)
        mBinding.searchCategoryRoot.tvSearchArtist.setOnClickListener(this)
        mBinding.searchCategoryRoot.tvSearchAlbum.setOnClickListener(this)
        mBinding.ivClearHistory.setOnClickListener(this)
        mBinding.editSearch.addTextChangedListener(object : TextChangedListener() {
            override fun afterTextChanged(s: Editable) {
                searchMusic(s.toString().trim(), mPosition)
            }
        })
        mBinding.smartisanControlBar.setOnClickListener { startPlayActivity() }
        // 底部控制Bar按钮监听
        mBinding.smartisanControlBar.setClickListener { clickFlag: Int ->
            LogUtil.d(TAG, clickFlag.toString())
            when (clickFlag) {
                Constant.NUMBER_ONE -> {
                    audioBinder!!.updateFavorite()
                    checkCurrentSongIsFavorite(mMusicBean, null, mBinding.smartisanControlBar)
                }

                Constant.NUMBER_TWO -> audioBinder!!.playPre()
                Constant.NUMBER_THREE -> switchPlayState()
                Constant.NUMBER_FOUR -> audioBinder!!.playNext()
                else -> {}
            }
        }
    }


    override fun onClick(v: View) {
        when (v.id) {
            R.id.iv_edit_clear -> {
                mBinding.editSearch.setText("")
                mAdapter?.clear()
            }

            R.id.iv_clear_history -> {
                LogUtil.d(TAG, "历史点击")
                // 清空历史记录
            }

            R.id.tv_search_all -> {
                switchListCategory(Constant.SEARCH_SONG_FOR_ALL)
            }

            R.id.tv_search_song -> {
                switchListCategory(Constant.SEARCH_SONG_FOR_NAME)
            }

            R.id.tv_search_album -> {
                switchListCategory(Constant.SEARCH_SONG_FOR_ALL)
            }

            R.id.tv_search_artist -> {
                switchListCategory(Constant.SEARCH_SONG_FOR_ARTIST)
            }

            R.id.tv_search_cancel -> {
                SoftKeybordUtil.showAndHintSoftInput(
                    mInputMethodManager, 1, InputMethodManager.RESULT_UNCHANGED_SHOWN
                )
                finish()
            }
        }
    }

    /**
     * @param searchKey 搜索关键字key 、
     * @param position 搜索类别：  11歌曲 、 12专辑 、 13 艺术家 、14 全部
     */
    private fun searchMusic(searchKey: String, position: Int) {
        mBinding.ivEditClear.visibility = if (searchKey.isEmpty()) View.GONE else View.VISIBLE
        mBinding.tvNoSearchResult.visibility = if (searchKey.isEmpty()) View.GONE else View.VISIBLE
        mBinding.flowlayout.visibility = if (searchKey.isEmpty()) View.VISIBLE else View.GONE

        mBinding.searchCategoryRoot.root.visibility =
            if (searchKey.isEmpty()) View.GONE else View.VISIBLE
        mBinding.groupSearch.visibility = if (searchKey.isEmpty()) View.VISIBLE else View.GONE
        if (searchKey.isNotEmpty()) {
            mViewModel.searchMusic(searchKey, position)
        } else {
            // 获取历史记录
            mViewModel.getHistory()
        }
    }

    override fun updateLyricsView(lyricsOK: Boolean, downMsg: String) {
        if (lyricsOK) {
            updateLyric()
        }
    }


    override fun updateCurrentPlayInfo(musicItem: MusicBean) {
        mMusicBean = musicItem
        setMusicInfo(musicItem)
        mBinding.smartisanControlBar.initAnimation()
        if (audioBinder != null) {
            setDuration()
            checkCurrentSongIsFavorite(mMusicBean, null, mBinding.smartisanControlBar)
            mBinding.smartisanControlBar.updatePlayBtnStatus(audioBinder!!.isPlaying)
            updateLyric()
        }
    }

    override fun updateCurrentPlayProgress() {
        if (audioBinder != null) {
            mBinding.smartisanControlBar.setSongProgress(audioBinder!!.progress)
        }
    }

    private fun setDuration() {
        val duration = audioBinder!!.duration
        mBinding.smartisanControlBar.setMaxProgress(duration)
    }

    private fun setMusicInfo(bean: MusicBean?) {
        if (bean != null) {
            mBinding.smartisanControlBar.visibility = View.VISIBLE
            val musicItem = TitleArtistUtil.getMusicBean(bean)
            mBinding.smartisanControlBar.setSongName(musicItem.title)
            mBinding.smartisanControlBar.setSingerName(musicItem.artist)
            mBinding.smartisanControlBar.setAlbumUrl(this, musicItem)
        }
        if (audioBinder != null) {
            mBinding.smartisanControlBar.updatePlayBtnStatus(audioBinder!!.isPlaying)
            mBinding.smartisanControlBar.initAnimation()
        } else {
            mBinding.smartisanControlBar.visibility = View.GONE
        }
    }

    private fun switchListCategory(position: Int) {
        mPosition = position
        val searchKey = mBinding.editSearch.text.trim().toString()
        mViewModel.searchMusic(searchKey, position)
        when (position) {
            Constant.SEARCH_SONG_FOR_NAME -> {
                setAllCategoryNotNormal()
                mBinding.searchCategoryRoot.tvSearchSong.setTextColor(ColorUtil.wihtle)
                mBinding.searchCategoryRoot.tvSearchSong.setBackgroundResource(R.drawable.btn_category_start_down_selector)
            }

            Constant.SEARCH_SONG_FOR_ALBUM -> {
                setAllCategoryNotNormal()
                mBinding.searchCategoryRoot.tvSearchAlbum.setTextColor(ColorUtil.wihtle)
                mBinding.searchCategoryRoot.tvSearchAlbum.setBackgroundResource(R.drawable.btn_category_middle_down_selector)
            }

            Constant.SEARCH_SONG_FOR_ARTIST -> {
                setAllCategoryNotNormal()
                mBinding.searchCategoryRoot.tvSearchArtist.setBackgroundResource(R.drawable.btn_category_middle_down_selector)
                mBinding.searchCategoryRoot.tvSearchArtist.setTextColor(ColorUtil.wihtle)
            }

            Constant.SEARCH_SONG_FOR_ALL -> {
                setAllCategoryNotNormal()
                mBinding.searchCategoryRoot.tvSearchAll.setTextColor(ColorUtil.wihtle)
                mBinding.searchCategoryRoot.tvSearchAll.setBackgroundResource(R.drawable.btn_category_end_down_selector)
            }
        }
    }

    private fun setAllCategoryNotNormal() {
        mBinding.searchCategoryRoot.tvSearchSong.setTextColor(ColorUtil.textName)
        mBinding.searchCategoryRoot.tvSearchSong.setBackgroundResource(R.drawable.btn_category_start_selector)
        mBinding.searchCategoryRoot.tvSearchAlbum.setTextColor(ColorUtil.textName)
        mBinding.searchCategoryRoot.tvSearchAlbum.setBackgroundResource(R.drawable.btn_category_middle_selector)
        mBinding.searchCategoryRoot.tvSearchArtist.setTextColor(ColorUtil.textName)
        mBinding.searchCategoryRoot.tvSearchArtist.setBackgroundResource(R.drawable.btn_category_middle_selector)
        mBinding.searchCategoryRoot.tvSearchAll.setTextColor(ColorUtil.textName)
        mBinding.searchCategoryRoot.tvSearchAll.setBackgroundResource(R.drawable.btn_category_end_selector)
    }

    override fun refreshBtnAndNotify(playStatus: Int) {
        when (playStatus) {
            0 -> {
                mBinding.smartisanControlBar.animatorOnResume(audioBinder!!.isPlaying)
                mBinding.smartisanControlBar.updatePlayBtnStatus(audioBinder!!.isPlaying)
            }

            1 -> {
                checkCurrentSongIsFavorite(mMusicBean, null, mBinding.smartisanControlBar)
            }

            2 -> {
                mBinding.smartisanControlBar.updatePlayBtnStatus(audioBinder!!.isPlaying)
                mBinding.smartisanControlBar.animatorOnPause()
            }

        }
    }


    private fun switchPlayState() {
        if (audioBinder!!.isPlaying) {
            audioBinder!!.pause()
        } else {
            audioBinder!!.start()
        }
        mBinding.smartisanControlBar.animatorOnResume(audioBinder!!.isPlaying)
        mBinding.smartisanControlBar.updatePlayBtnStatus(audioBinder!!.isPlaying)
    }


    override fun startMusicServiceFlag(
        position: Int, pageType: Int, conditon: String
    ) {
        val condition = mBinding.editSearch.text.toString().trim()
        val intent = Intent(this, MusicPlayService::class.java)
        intent.putExtra(Constant.PAGE_TYPE, pageType)
        intent.putExtra(Constant.CONDITION, condition)
        intent.putExtra(Constant.POSITION, position)
        startService(intent)
    }

    override fun startMusicService(position: Int, mPageType: Int) {}
    override fun onOpenMusicPlayDialogFag() {}
    override fun onPause() {
        super.onPause()
        mBinding.smartisanControlBar.animatorOnPause()
    }

    override fun onDestroy() {
        super.onDestroy()
        if (audioBinder != null) {
            audioBinder = null
        }
    }

    private fun updateLyric() {
        val lyricList = LyricsUtil.getLyricList(mMusicBean)
        disposableQqLyric()
        if (mQqLyricsDisposable == null) {
            mQqLyricsDisposable =
                Observable.interval(0, 2800, TimeUnit.MICROSECONDS).subscribeOn(Schedulers.io())
                    .observeOn(AndroidSchedulers.mainThread()).subscribe {
                        if (lyricList.size > 1 && lyricsFlag < lyricList.size) {
                            //通过集合，播放过的歌词就从集合中删除
                            val lyrBean = lyricList[lyricsFlag]
                            val content = lyrBean.content
                            val progress = audioBinder!!.progress
                            val startTime = lyrBean.startTime
                            if (progress > startTime) {
                                mBinding.smartisanControlBar.setSingerName(content)
                                lyricsFlag++
                            }
                        }
                    }
        }
    }

    override fun click(songName: String) {
        mBinding.editSearch.setText(songName)
        mBinding.editSearch.setSelection(songName.length)
    }


    override fun finish() {
        super.finish()
        overridePendingTransition(0, R.anim.dialog_push_out)
    }
}